diff -u -N -r linux-4.6.2/drivers/mmc/host/Kconfig linux-4.6.2-clean/drivers/mmc/host/Kconfig
--- linux-4.6.2/drivers/mmc/host/Kconfig	2016-06-08 02:23:53.000000000 +0100
+++ linux-4.6.2-clean/drivers/mmc/host/Kconfig	2017-03-30 17:17:51.238686000 +0100
@@ -164,6 +164,17 @@
 
 	  If unsure, say N.
 
+config MMC_SDHCI_OF_MINION
+	tristate "SDHCI OF support for the lowRISC Minion SDHCI controllers"
+	depends on MMC_SDHCI_PLTFM
+	depends on RISCV
+	select MMC_SDHCI_IO_ACCESSORS
+	help
+	  This selects the Secure Digital Host Controller Interface (SDHCI)
+	  found in the lowRISC Minion FPGA.
+
+	  If you don't know what this is, just say N.
+
 config MMC_SDHCI_CNS3XXX
 	tristate "SDHCI support on the Cavium Networks CNS3xxx SoC"
 	depends on ARCH_CNS3XXX
diff -u -N -r linux-4.6.2/drivers/mmc/host/Makefile linux-4.6.2-clean/drivers/mmc/host/Makefile
--- linux-4.6.2/drivers/mmc/host/Makefile	2016-06-08 02:23:53.000000000 +0100
+++ linux-4.6.2-clean/drivers/mmc/host/Makefile	2017-03-30 17:18:10.990054000 +0100
@@ -70,6 +70,7 @@
 obj-$(CONFIG_MMC_SDHCI_OF_AT91)		+= sdhci-of-at91.o
 obj-$(CONFIG_MMC_SDHCI_OF_ESDHC)	+= sdhci-of-esdhc.o
 obj-$(CONFIG_MMC_SDHCI_OF_HLWD)		+= sdhci-of-hlwd.o
+obj-$(CONFIG_MMC_SDHCI_OF_MINION)	+= sdhci-of-minion.o
 obj-$(CONFIG_MMC_SDHCI_BCM_KONA)	+= sdhci-bcm-kona.o
 obj-$(CONFIG_MMC_SDHCI_BCM2835)		+= sdhci-bcm2835.o
 obj-$(CONFIG_MMC_SDHCI_IPROC)		+= sdhci-iproc.o
diff -u -N -r linux-4.6.2/drivers/mmc/host/sdhci.c linux-4.6.2-clean/drivers/mmc/host/sdhci.c
--- linux-4.6.2/drivers/mmc/host/sdhci.c	2016-06-08 02:23:53.000000000 +0100
+++ linux-4.6.2-clean/drivers/mmc/host/sdhci.c	2017-04-01 19:09:58.760645673 +0100
@@ -2126,6 +2126,20 @@
 	spin_unlock_irqrestore(&host->lock, flags);
 }
 
+#ifdef CONFIG_MMC_SDHCI_OF_MINION
+static int sdhci_multi_io_quirk(struct mmc_card *card, unsigned int direction, int blk_size)
+{
+        /* This controller can't do multiblock reads due to hw bugs */
+        if (direction == MMC_DATA_READ)
+                return 1;
+
+        if (direction == MMC_DATA_WRITE)
+                return 1;
+
+        return blk_size;
+}
+#endif
+
 static const struct mmc_host_ops sdhci_ops = {
 	.request	= sdhci_request,
 	.post_req	= sdhci_post_req,
@@ -2141,6 +2155,9 @@
 	.select_drive_strength		= sdhci_select_drive_strength,
 	.card_event			= sdhci_card_event,
 	.card_busy	= sdhci_card_busy,
+#ifdef CONFIG_MMC_SDHCI_OF_MINION
+	.multi_io_quirk = sdhci_multi_io_quirk
+#endif
 };
 
 /*****************************************************************************\
